common/wireguard/src/ip_pool.rs

```diff
@@ -0,0 +1,202 @@
+#[derive(Clone)]
+pub struct IpPool {
+    allocations: Arc<RwLock<HashMap<IpPair, AllocationState>>>,
+}
+
+impl IpPool {
+    /// Create a new IP pool from IPv4 and IPv6 CIDR ranges
+    ///
+    /// # Arguments
+    /// * `ipv4_network` - Base IPv4 address for the pool
+    /// * `ipv4_prefix` - CIDR prefix length for IPv4 (e.g., 16 for /16)
+    /// * `ipv6_network` - Base IPv6 address for the pool
+    /// * `ipv6_prefix` - CIDR prefix length for IPv6 (e.g., 112 for /112)
+    ///
+    /// # Errors
+    /// Returns error if CIDR ranges are invalid
+    pub fn new(
+        ipv4_network: Ipv4Addr,
+        ipv4_prefix: u8,
+        ipv6_network: Ipv6Addr,
+        ipv6_prefix: u8,
+    ) -> Result<Self, IpPoolError> {
+        let ipv4_net = IpNetwork::new(ipv4_network.into(), ipv4_prefix)?;
+        let ipv6_net = IpNetwork::new(ipv6_network.into(), ipv6_prefix)?;
+
+        // Build initial pool with all IPs marked as free
+        let mut allocations = HashMap::new();
+
+        // Collect IPv4 and IPv6 addresses into vectors for pairing
+        let ipv4_addrs: Vec<Ipv4Addr> = ipv4_net
+            .iter()
+            .filter_map(|ip| {
+                if let IpAddr::V4(v4) = ip {
+                    Some(v4)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let ipv6_addrs: Vec<Ipv6Addr> = ipv6_net
+            .iter()
+            .filter_map(|ip| {
+                if let IpAddr::V6(v6) = ip {
+                    Some(v6)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Create IpPairs by matching IPv4 and IPv6 addresses
+        // Use the minimum length to avoid index out of bounds
+        let pair_count = ipv4_addrs.len().min(ipv6_addrs.len());
+        for i in 0..pair_count {
+            let pair = IpPair::new(ipv4_addrs[i], ipv6_addrs[i]);
+            allocations.insert(pair, AllocationState::Free);
+        }
+
+        tracing::info!(
+            "Initialized IP pool with {} address pairs from {}/{} and {}/{}",
+            allocations.len(),
+            ipv4_network,
+            ipv4_prefix,
+            ipv6_network,
+            ipv6_prefix
+        );
+
+        Ok(IpPool {
+            allocations: Arc::new(RwLock::new(allocations)),
+        })
+    }
+
+    /// Allocate a free IP pair from the pool
+    ///
+    /// Randomly selects an available IP pair and marks it as allocated.
+    ///
+    /// # Errors
+    /// Returns `IpPoolError::NoFreeIp` if no IPs are available
+    pub async fn allocate(&self) -> Result<IpPair, IpPoolError> {
+        let mut pool = self.allocations.write().await;
+
+        // Find a free IP and allocate it
+        let free_ip = pool
+            .iter_mut()
+            .filter(|(_, state)| matches!(state, AllocationState::Free))
+            .choose(&mut rand::thread_rng())
```

Comment:
wouldn't this form of allocation be quite costly as we'd have to iterate the whole map? why not
1) have two structures, one for allocated and one for free IpPair
2) allocate addresses sequentially from the free pool